content/js2/blocks/reference/index.md

```diff
@@ -0,0 +1,130 @@
+If we mutate `list` then we're mutating the same list that `copy` points to.
+
+So the console output is the same.
+
+{{</tab>}}
+
+{{<tab name=" 🧠  Explain">}}
+
+```js {linenos=table,hl_lines=["4"],linenostart=1}
+const salaries = [10, 20, 30, 40, 60, 80, 80];
+const median = calculateMedian(salaries);
+
+console.log(salaries, "<--- salaries input before we call calculateMean");
+const mean = calculateMean(salaries);
+
+console.log(`The median salary is ${median}`);
+console.log(`The mean salary is ${mean}`);
+```
+
+In the example above, `salaries` is assigned a reference on the first line.
+Explain why `calculateMedian` and `calculateMean` both get access to the same array.
+
+{{</tab>}}
+
+{{</tabs>}}
+
+### Shared reference
+
+We can also check these variables share the same reference.
+
+```js
+const list = [10, 20, 30];
+const copy = list;
+
+console.log(list === copy); // evaluates to true
+```
+
+If we're comparing 2 array variables with `===`, then it will evaluate to `true` only if they have the **same reference**. `===` is comparing the references to the arrays, not the arrays themselves.
+
+### Value vs reference
+
+In JavaScript, arrays and objects are reference types: everything else is a value type.
+{{<note title="Passing by value" type="note">}}
+
+As strings are value types, they are **passed by value**. Passed by value means that every time you assign a value to a variable then a copy of that value is made.
+
+{{</note>}}
+
+Use the tabs below to compare the effects of **passing by reference** and **passing by value**.
+There are two different but similar implementations of `pluralise` - a function that appends an `s` to the end of its input.
+
+{{<tabs>}}
+
+{{<tab name="pluralise: array">}}
+
+Here `pluralise` is **passed an array by reference**.
+
+`lettersInAnArray` is passed by reference. `pluralise`'s modification is visible here, because the same underlying storage was modified.
+
+Step through the code to observe this behaviour:
+
+<iframe title="pluralise-string" width="1100" height="500" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=function%20pluralise%28arr%29%20%7B%0A%20%20arr.push%28%22s%22%29%3B%0A%7D%0Aconst%20lettersInAnArray%20%3D%20%5B%22c%22,%22a%22,%22t%22%5D%3B%0Apluralise%28lettersInAnArray%29%3B%0A%0Aconsole.assert%28lettersInAnArray.length%20%3D%3D%3D%204%29%3B%0Aconsole.assert%28lettersInAnArray%5B3%5D%20%3D%3D%3D%20%22s%22%29%3B&codeDivHeight=400&codeDivWidth=600&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>
+{{</tab>}}
+
+{{<tab name="pluralise: string">}}
+
+Here `pluralise` is **passed a string by value**.
+
+This means a copy of `string`'s value is passed to `pluralise` in the second tab. `pluralise`'s reassignment is not visible here, because a copy was made just for the function before the value was modified.
+
+Step through the code to observe this behaviour:
```

Comment:
The code in this one is buggy - the assertions are:

```js
console.assert(string.length === 1);
console.assert(string === "a");
```

where they should be:
```js
console.assert(string.length === 3);
console.assert(string === "cat");
```